// @hyperlane-xyz/core/dist/factories/contracts/isms/hook/CrossChainEnabled__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrossChainEnabled__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "actual",
                type: "address",
            },
            {
                internalType: "address",
                name: "expected",
                type: "address",
            },
        ],
        name: "InvalidCrossChainSender",
        type: "error",
    },
    {
        inputs: [],
        name: "NotCrossChainCall",
        type: "error",
    },
];
class CrossChainEnabled__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.CrossChainEnabled__factory = CrossChainEnabled__factory;
CrossChainEnabled__factory.abi = _abi;
